### ISSUE TITLE
subunit reporter test failures

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4608]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)| @mithrandi reported|
|-|-|
|Trac ID|trac#4608|
|Type|defect|
|Created|2010-07-21 00:33:42Z|
|Branch|https://github.com/twisted/twisted/tree/robust-subunit-tests-4608-2|

I'm not sure if I have a version of subunit that is too old, or something (I have python-subunit 0.0.6-1 from debian unstable), but I'm getting the following test failures in trunk@HEAD:

```
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/home/mithrandi/code/Twisted/trunk/twisted/trial/test/test_reporter.py", line 1228, in test_subunitWithoutAddExpectedFailureInstalled
    self.assertEquals(successOutput, expectedFailureOutput)
twisted.trial.unittest.FailTest: not equal:
a = 'successful: twisted.trial.test.sample.FooTest.test_foo\n'
b = ''


twisted.trial.test.test_reporter.TestSubunitReporter.test_subunitWithoutAddExpectedFailureInstalled
===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/home/mithrandi/code/Twisted/trunk/twisted/trial/test/test_reporter.py", line 1251, in test_subunitWithoutAddSkipInstalled
    self.assertEquals(successOutput, skipOutput)
twisted.trial.unittest.FailTest: not equal:
a = 'successful: twisted.trial.test.sample.FooTest.test_foo\n'
b = ''


twisted.trial.test.test_reporter.TestSubunitReporter.test_subunitWithoutAddSkipInstalled
```

<details><summary>Searchable metadata</summary>

```
trac-id__4608 4608
type__defect defect
reporter__mithrandi mithrandi
priority__normal normal
milestone__ 
branch__branches_robust_subunit_tests_4608_2 branches/robust-subunit-tests-4608-2
branch_author__jml jml
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1279672422000000 1279672422000000
changetime__1285164925000000 1285164925000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

The tests pass on my system with python-subunit 0.0.2-1 from Ubuntu Karmic.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml set status to `assigned`|
|-|-|

I can reproduce this bug with subunit 0.0.6 and subunit trunk r132.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

A little experimentation shows that the test failure is introduce with subunit trunk r130. That revision changes `subunit.TestProtocolClient` to inherit from `testtools.TestResult`, which provides `addExpectedFailure` and `addSkip`. Since those two methods exist on the base class, the tests in trial (`test_subunitWithoutAddExpectedFailureInstalled` and `test_subunitWithoutAddSkipInstalled`), delete only subunit's override of these methods rather than deleting the methods entirely.

Since the tests are designed to exercise SubunitReporter's ability to handle old versions of subunit, I have changed the tests to delete the methods from `TestProtocolClient` and all of its base classes. Code forthcoming.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner<br>@jml set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @jml|
|-|-|

There's a trivial conflict with trunk.  More importantly, though, the changes don't seem to amount to much.  There's a new method, `removeMethod`, but nothing uses it.  So I don't see how this would help.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

(In [30052]) Branching to 'robust-subunit-tests-4608-2'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml commented|
|-|-|

The "trivial" conflict with trunk is Subversion hiding the more interesting changes in the branch. I've "rebranched", re-applied the changes and verified the diff with Combinator.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @jml|
|-|-|

Looks good to me, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml set status to `closed`|
|-|-|

(In [30072]) Merge robust-subunit-tests-4608-2

 * Author: jml
 * Reviewers: exarkun, therve
 * Fixes [#4608](https://github.com/twisted/twisted/issues/4608)

Make the subunit reporter tests more robust against varying versions
of subunit.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
